File: skill-markets/agent-dev-control-kit/scaffolds/python/files/guards/api-contract-guard.py
#!/usr/bin/env python3
"""api-contract-guard.py — verify that every service module exposes callable symbols.

Exit 0 on pass, 1 on fail.
"""

from __future__ import annotations

import ast
import sys
from pathlib import Path

ROOT = Path.cwd()
SERVICES_DIR = ROOT / "src" / "services"
SCHEMAS_DIR = ROOT / "schemas" / "api"


def list_modules(d: Path) -> list[Path]:
    if not d.exists():
        return []
    return [p for p in d.glob("*.py") if not p.name.startswith("_")]


def extract_top_level_funcs(path: Path) -> list[str]:
    tree = ast.parse(path.read_text(encoding="utf-8"))
    return [n.name for n in tree.body if isinstance(n, (ast.FunctionDef, ast.AsyncFunctionDef))]


def main() -> int:
    modules = list_modules(SERVICES_DIR)
    if not modules:
        print("[api-contract-guard] No src/services/ — skipped.")
        return 0

    failures = 0
    for m in modules:
        funcs = extract_top_level_funcs(m)
        print(f"[api-contract-guard] {m.name}: funcs = [{', '.join(funcs) or 'none'}]")
        if not funcs:
            print(f"[api-contract-guard] FAIL: {m.name} has no top-level functions")
            failures += 1

    schemas = list(SCHEMAS_DIR.glob("*.json")) if SCHEMAS_DIR.exists() else []
    print(f"[api-contract-guard] schemas: {len(schemas)} found")

    if failures:
        print(f"[api-contract-guard] {failures} failure(s)")
        return 1
    print("[api-contract-guard] PASSED")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: skill-markets/agent-dev-control-kit/scaffolds/nodejs/files/guards/api-contract-guard.mjs
#!/usr/bin/env node
// api-contract-guard.mjs
// Verifies that every service module exports functions matching the API schema.
// Exit 0 = pass, 1 = fail.

import { readdirSync, readFileSync, existsSync } from 'node:fs';
import { join, resolve } from 'node:path';

const ROOT = process.cwd();
const SERVICES_DIR = resolve(ROOT, 'src/services');
const SCHEMAS_DIR = resolve(ROOT, 'schemas/api');

function listFiles(dir) {
  if (!existsSync(dir)) return [];
  return readdirSync(dir).filter((f) => f.endsWith('.js') || f.endsWith('.mjs'));
}

function extractExports(filePath) {
  const src = readFileSync(filePath, 'utf8');
  const matches = [...src.matchAll(/export\s+(?:async\s+)?function\s+(\w+)/g)];
  return matches.map((m) => m[1]);
}

function main() {
  if (!existsSync(SERVICES_DIR)) {
    console.log('[api-contract-guard] No src/services/ found — skipped.');
    return 0;
  }

  const serviceFiles = listFiles(SERVICES_DIR);
  const schemaFiles = listFiles(SCHEMAS_DIR).filter((f) => f.endsWith('.json'));

  if (serviceFiles.length === 0) {
    console.log('[api-contract-guard] No services found — skipped.');
    return 0;
  }

  let failures = 0;
  for (const svc of serviceFiles) {
    const exports = extractExports(join(SERVICES_DIR, svc));
    console.log(`[api-contract-guard] ${svc}: exports = [${exports.join(', ') || 'none'}]`);
    if (exports.length === 0) {
      console.error(`[api-contract-guard] FAIL: ${svc} has no exported functions`);
      failures += 1;
    }
  }

  console.log(`[api-contract-guard] schemas: ${schemaFiles.length} found`);

  if (failures > 0) {
    console.error(`[api-contract-guard] ${failures} failure(s)`);
    return 1;
  }
  console.log('[api-contract-guard] PASSED');
  return 0;
}

process.exit(main());